### PR TITLE
feat: adding groups adr

### PR DESCRIPTION
**Pull request by kiram15**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:27]: just for completeness sake can we make sure fields are labeled with all attributes? (ie a NOT NULL, default values, typing etc)

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:15]: extra space?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:56]: spacing?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:34]: group_uuid probably should be part of the path, not a query param

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:36]: `Accepted Query Params` > `Inputs` IMO

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:43]: you can probably just label this `/learners/` with the path below

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:49]: "Verifies provided learner_ids belong to members of the group"?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:53]: an example response body would be nice

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:29]: not sure this is needed, it would also be the list endpoint for the view

Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:29]: (thus a part of CRUD)

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:43]: GET `/learners/`

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:41]: Do you think an output and/or example response body would be helpful?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:57]: POST `/assign_learners/`

Also I'm curious if we want to do `assign_learners` or `assign-learners`

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:64]: might be worth returning the membership uuid on success

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:72]: not assign, remind

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:81]: Thinking about what we should do for partial successes, what responses should the consumer expect if they provide both learners that exist in the group and learners that don't?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:91]: make this clear that it's a char selection? Similar to
```
**EnterpriseFulfillmentSource**

\------------------------------\

- uuid, created, modified, history (boilerplate)
- fulfillment_type (NOT NULL, char selection: (`license`, `learner_credit`, `coupon_code`))
- enterprise_course_entitlement (NOT NULL, FK to EnterpriseCourseEntitlement)
- enterprise_course_enrollment (NOT NULL, FK to EnterpriseCourseEnrollment)
- is_revoked (Default False, Bool)
```

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:20]: if we wanted, we can maybe add a section here on how systems will interact, part of the intention for this table will be to add further filtering on the `/enterprise-learner/` endpoint - this could also maybe go into a consequences section?

**Review by alex-sheehan-edx — COMMENTED** [review]
Good start, some initial cleanup comments, I would maybe preview your rst document with the vscode plugin or some webtool to make sure your formatting is what you think it is. Also would consider writing a `consequences` section with the intended effects this new architecture will have on the platform.

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:92]: should probably be Nullable?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (kiram15) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:34]: No because in this case if they don't want to search by group uuid, they might want to search by enterprise uuid which would have multiple groups

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:36]: hmm actually if we have `list` documented separately from `fetch` I think this is a more appropriate `list` request response not a fetch?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:138]: [Question] do we want to return the membership uuids of the learners removed?

Comment (kiram15) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:138]: I don't know I don't really see the point of it? What use could we have of them?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:37]: should group be an array of objects or an object?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:135]: it gives consumers more information on the exact records that were modified as a result of their request but I don't think it's fully necessary for the operations we're looking to perform for MVP

**Review by alex-sheehan-edx — APPROVED** [review]
lgtm! idk if you want someone else to review before merging. I thiiiiink we'll want a change log and version bump? Don't forget to update the status to accepted 😄

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (johnnagro) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:11]: Groups are a feature of an enterprise more than a feature of learner-credit. Groups will be used for lots of things like access control, analytics segmentation, bulk actions throughout the enterprise admin experience, etc. I'd caution us against pigeonholing groups into one thing like "...collection of enterprise customer users connected to a policy/subsidy".

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (johnnagro) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:18]: ditto, its more than just control and distribute funds. I'd add things like analytics segmentation, bulk actions, access control, as well as budgets. They are not tied to the policy as much as policies can make use of groups.

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (johnnagro) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:26]: policies were actually designed to have a group_uuid - groups should not have a policy_uuid

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (johnnagro) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:122]: remind them means what?

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (johnnagro) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:141]: imho we shouldnt include any of this complexity in V1. learners are assigned to a group without having to accept that invitation (in the mocks ive seen thus far). this table also doesnt really need its own PK - the group_uuid, lms_user_id pair is enough to address into a specific membership - unless django ORM requires this but if its just a join table (drop the remind, status stuff) this shouldnt need to be too complicated.

**Review by johnnagro — CHANGES_REQUESTED** [review]
groups will have a wider scope of value, wider than just learner credit. as such imho we should tweak this design to be less learner-credit specific.

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:122]: remind to accept the group invitation

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [reply docs/decisions/0013-custom-groups-for-enterprise-learners.rst:141]: We should talk to product because it was my assumption that pending group status was a requirement for MVP

File: docs/decisions/0013-custom-groups-for-enterprise-learners.rst
Comment (alex-sheehan-edx) [inline docs/decisions/0013-custom-groups-for-enterprise-learners.rst:132]: are we going to have a pending customer user id here as well?

**Review by johnnagro — APPROVED** [review]

**Pull request merged**
